libhb/work.c

```diff
@@ -1229,169 +1229,69 @@ static int sanitize_audio(hb_job_t *job)
-                }
-                job->qsv.decode = 0;
-            }
+            hb_error( "sanitize_qsv: qsv ctx alloc failed" );
+            return 1;
         }
+        job->qsv.ctx->hb_dec_qsv_frames_ctx = av_mallocz(sizeof(HBQSVFramesContext));
+        if(!job->qsv.ctx->hb_dec_qsv_frames_ctx)
+        {
+            hb_error( "sanitize_qsv: HBQSVFramesContext dec alloc failed" );
+            return 1;
+        }
+        hb_qsv_add_context_usage(job->qsv.ctx, 0);
     }
 
-    /*
-     * When QSV's VPP is used for filtering, not all CPU filters
-     * are supported, so we need to do a little extra setup here.
-     */
+    int i = 0;
+    int num_cpu_filters = 0;
     if (job->vcodec & HB_VCODEC_QSV_MASK)
     {
-        int vpp_settings[7];
-        int num_cpu_filters = 0;
-        hb_filter_object_t *filter;
-        // default values for VPP filter
-        vpp_settings[0] = job->title->geometry.width;
-        vpp_settings[1] = job->title->geometry.height;
-        vpp_settings[2] = job->title->crop[0];
-        vpp_settings[3] = job->title->crop[1];
-        vpp_settings[4] = job->title->crop[2];
-        vpp_settings[5] = job->title->crop[3];
-        vpp_settings[6] = 0; // deinterlace: off
         if (job->list_filter != NULL && hb_list_count(job->list_filter) > 0)
         {
-            while (hb_list_count(job->list_filter) > num_cpu_filters)
+            for (i = 0; i < hb_list_count(job->list_filter); i++)
             {
-                filter = hb_list_item(job->list_filter, num_cpu_filters);
+                hb_filter_object_t *filter = hb_list_item(job->list_filter, i);
+
                 switch (filter->id)
                 {
                     // cropping and scaling always done via VPP filter
                     case HB_FILTER_CROP_SCALE:
-                        hb_dict_extract_int(&vpp_settings[0], filter->settings,
-                                            "width");
-                        hb_dict_extract_int(&vpp_settings[1], filter->settings,
-                                            "height");
-                        hb_dict_extract_int(&vpp_settings[2], filter->settings,
-                                            "crop-top");
-                        hb_dict_extract_int(&vpp_settings[3], filter->settings,
-                                            "crop-bottom");
-                        hb_dict_extract_int(&vpp_settings[4], filter->settings,
-                                            "crop-left");
-                        hb_dict_extract_int(&vpp_settings[5], filter->settings,
-                                            "crop-right");
-
-                        hb_list_rem(job->list_filter, filter);
-                        hb_filter_close(&filter);
                         break;
 
-                    // pick VPP or CPU deinterlace depending on settings
                     case HB_FILTER_DEINTERLACE:
-                    {
```

Comment:
You removed the check for QSV deinterlace filter.  Is that intentional?  QSV deinterlace filter is now counted as a cpu filter.  It appears that the way qsv buffers are copied in and out of system memory has changed so maybe this is no longer relevant?

Reply:
This part of code was completely commented before. 

Deinterlace is expected to be HW QSV accelerated filter as HW crop/scale via FFMPEG in the future. Obviously this is next step. 

It is separate topic but for now deinterlace detection is enabled by default in pipeline and this could be bottleneck for case where the progressive content is widely used. Copying in and out system memory just to make deinterlace detection does look efficient especially when QSV is used slowdown whole pipeline.

Reply:
@j45 deinterlace will be offloaded to use libavfilter/vf_deinterlace_qsv.c from ffmpeg, 
not in this patch though.